File: java/com/google/copybara/transform/VerifyMatch.java
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.transform;

import static com.google.common.base.Preconditions.checkNotNull;

import com.google.common.base.MoreObjects;
import com.google.common.base.Splitter;
import com.google.common.collect.Iterables;
import com.google.copybara.LocalParallelizer;
import com.google.copybara.LocalParallelizer.TransformFunc;
import com.google.copybara.TransformWork;
import com.google.copybara.Transformation;
import com.google.copybara.TransformationStatus;
import com.google.copybara.exception.ValidationException;
import com.google.copybara.treestate.TreeState.FileState;
import com.google.copybara.util.Glob;
import com.google.re2j.Matcher;
import com.google.re2j.Pattern;
import com.google.re2j.PatternSyntaxException;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import javax.annotation.Nullable;
import net.starlark.java.eval.EvalException;
import net.starlark.java.eval.Starlark;
import net.starlark.java.syntax.Location;

/**
 * A source code pseudo-transformation which verifies that all specified files satisfy a RegEx.
 * Does not actually transform any code, but will throw errors on failure. Not applied in reversals.
 */
public final class VerifyMatch implements Transformation {

  private final Pattern pattern;
  private final boolean verifyNoMatch;
  private final boolean alsoOnReversal;
  private final Glob fileMatcherBuilder;
  private final LocalParallelizer parallelizer;
  private final Location location;
  private final Optional<String> failureMessage;

  private VerifyMatch(Pattern pattern, boolean verifyNoMatch, boolean alsoOnReversal,
      Glob fileMatcherBuilder, Optional<String> failureMessage, LocalParallelizer parallelizer,
      Location location) {
    this.pattern = checkNotNull(pattern);
    this.verifyNoMatch = verifyNoMatch;
    this.alsoOnReversal = alsoOnReversal;
    this.fileMatcherBuilder = checkNotNull(fileMatcherBuilder);
    this.failureMessage = checkNotNull(failureMessage);

    this.parallelizer = parallelizer;
    this.location = checkNotNull(location);
  }

  @Override
  public String toString() {
    return MoreObjects.toStringHelper(this)
        .add("Pattern", pattern)
        .add("verifyNoMatch", verifyNoMatch)
        .add("also_on_reversal", alsoOnReversal)
        .add("path", fileMatcherBuilder)
        .toString();
  }

  @Override
  public TransformationStatus transform(TransformWork work)
      throws IOException, ValidationException {
    Path checkoutDir = work.getCheckoutDir();
    Iterable<FileState> files = work.getTreeState().find(
        fileMatcherBuilder.relativeTo(checkoutDir));

    Iterable<String> errors = Iterables.concat(
        parallelizer.run(files, new BatchRun(work.getCheckoutDir())));

    int size = 0;
    for (String error : errors) {
      size++;
      work.getConsole().error(String.format("Error validating '%s': %s", describe(), error));
    }
    work.getTreeState().notifyNoChange();

    ValidationException.checkCondition(
        size == 0,
        "%d file(s) failed the validation of %s, located at %s.", size, describe(), location);

    return TransformationStatus.success();
  }

  private class BatchRun implements TransformFunc<FileState, List<String>> {

    private final Path checkoutDir;

    private BatchRun(Path checkoutDir) {
      this.checkoutDir = checkNotNull(checkoutDir);
    }

    @Override
    public List<String> run(Iterable<FileState> files)
        throws IOException, ValidationException {
      List<String> errors = new ArrayList<>();
      // TODO(malcon): Remove reconstructing pattern once RE2J doesn't synchronize on matching.
      Pattern batchPattern = Pattern.compile(pattern.pattern(), pattern.flags());
      for (FileState file : files) {
        if (Files.isSymbolicLink(file.getPath())) {
          continue;
        }
        String originalFileContent = Files.readString(file.getPath());
        Matcher matcher = batchPattern.matcher(originalFileContent);
        if (verifyNoMatch == matcher.find()) {
          String error = checkoutDir.relativize(file.getPath()).toString();
          if (verifyNoMatch) {
            int line = Splitter.on('\n')
                .splitToList(originalFileContent.substring(0, matcher.start())).size();
            error += String.format(" - Unexpected match found at line %d - '%s'.\n",
                line, matcher.group());
          } else {
            error += " - Expected string was not present.\n";
          }
          if (failureMessage.isPresent()) {
            error += failureMessage.get() + "\n";
          }
          errors.add(error);
        }
      }
      return errors;
    }
  }

  @Override
  public String describe() {
    return String.format("verify_match '%s'", pattern);
  }

  @Override
  public Location location() {
    return location;
  }

  @Override
  public Transformation reverse() {
    if (alsoOnReversal) {
      return new ExplicitReversal(this, this);
    }
    return new ExplicitReversal(IntentionalNoop.INSTANCE, this);
  }

  public static VerifyMatch create(Location location, String regEx, Glob paths,
      boolean verifyNoMatch, boolean alsoOnReversal, @Nullable String failureMessage,
      LocalParallelizer parallelizer) throws EvalException {
    Pattern parsed;
    try {
      parsed = Pattern.compile(regEx, Pattern.MULTILINE);
    } catch (PatternSyntaxException ex) {
      throw Starlark.errorf("Regex '%s' is invalid: %s", regEx, ex.getMessage());
    }
    return new VerifyMatch(parsed, verifyNoMatch, alsoOnReversal, paths,
        Optional.ofNullable(failureMessage), parallelizer, location);
  }
}

File: src/Copybara.Core/Transform/VerifyMatch.cs
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Text;
using System.Text.RegularExpressions;
using Copybara.Common;
using Copybara.Exceptions;
using Copybara.TreeState;
using Copybara.Util;
using Starlark.Syntax;
using StarlarkRt = Starlark.Eval.Starlark;
using FileState = Copybara.TreeState.TreeState.FileState;

namespace Copybara.Transform;

/// <summary>
/// A source code pseudo-transformation which verifies that all specified files satisfy a RegEx.
/// Does not actually transform any code, but will throw errors on failure. Not applied in reversals.
/// </summary>
public sealed class VerifyMatch : ITransformation
{
    private readonly Regex _pattern;
    private readonly bool _verifyNoMatch;
    private readonly bool _alsoOnReversal;
    private readonly Glob _fileMatcherBuilder;
    private readonly LocalParallelizer _parallelizer;
    private readonly Location _location;
    private readonly string? _failureMessage;

    private VerifyMatch(
        Regex pattern,
        bool verifyNoMatch,
        bool alsoOnReversal,
        Glob fileMatcherBuilder,
        string? failureMessage,
        LocalParallelizer parallelizer,
        Location location)
    {
        _pattern = Preconditions.CheckNotNull(pattern);
        _verifyNoMatch = verifyNoMatch;
        _alsoOnReversal = alsoOnReversal;
        _fileMatcherBuilder = Preconditions.CheckNotNull(fileMatcherBuilder);
        _failureMessage = failureMessage;
        _parallelizer = parallelizer;
        _location = Preconditions.CheckNotNull(location);
    }

    public override string ToString() =>
        $"VerifyMatch{{Pattern={_pattern}, verifyNoMatch={_verifyNoMatch},"
        + $" also_on_reversal={_alsoOnReversal}, path={_fileMatcherBuilder}}}";

    public TransformationStatus Transform(TransformWork work)
    {
        string checkoutDir = work.GetCheckoutDir();
        var files = work.GetTreeState().Find(_fileMatcherBuilder.RelativeTo(checkoutDir)).ToList();

        var errors = _parallelizer
            .Run(files, new BatchRun(this, work.GetCheckoutDir()))
            .SelectMany(e => e)
            .ToList();

        int size = 0;
        foreach (string error in errors)
        {
            size++;
            work.GetConsole().Error($"Error validating '{Describe()}': {error}");
        }
        work.GetTreeState().NotifyNoChange();

        ValidationException.CheckCondition(
            size == 0,
            "{0} file(s) failed the validation of {1}, located at {2}.", size, Describe(), _location);

        return TransformationStatus.Success();
    }

    private sealed class BatchRun : LocalParallelizer.TransformFunc<FileState, List<string>>
    {
        private readonly VerifyMatch _owner;
        private readonly string _checkoutDir;

        public BatchRun(VerifyMatch owner, string checkoutDir)
        {
            _owner = owner;
            _checkoutDir = Preconditions.CheckNotNull(checkoutDir);
        }

        public List<string> Run(IEnumerable<FileState> files)
        {
            var errors = new List<string>();
            var batchPattern = new Regex(_owner._pattern.ToString(), _owner._pattern.Options);
            foreach (FileState file in files)
            {
                var fileInfo = new FileInfo(file.GetPath());
                if (fileInfo.LinkTarget != null)
                {
                    continue;
                }
                string originalFileContent =
                    Encoding.UTF8.GetString(File.ReadAllBytes(file.GetPath()));
                Match matcher = batchPattern.Match(originalFileContent);
                if (_owner._verifyNoMatch == matcher.Success)
                {
                    string error = PathOps.Relativize(_checkoutDir, file.GetPath());
                    if (_owner._verifyNoMatch)
                    {
                        int line = originalFileContent
                            .Substring(0, matcher.Index)
                            .Split('\n')
                            .Length;
                        error += string.Format(
                            " - Unexpected match found at line {0} - '{1}'.\n", line, matcher.Value);
                    }
                    else
                    {
                        error += " - Expected string was not present.\n";
                    }
                    if (_owner._failureMessage != null)
                    {
                        error += _owner._failureMessage + "\n";
                    }
                    errors.Add(error);
                }
            }
            return errors;
        }
    }

    public string Describe() => $"verify_match '{_pattern}'";

    public Location Location() => _location;

    public ITransformation Reverse()
    {
        if (_alsoOnReversal)
        {
            return new ExplicitReversal(this, this);
        }
        return new ExplicitReversal(IntentionalNoop.Instance, this);
    }

    public static VerifyMatch Create(
        Location location,
        string regEx,
        Glob paths,
        bool verifyNoMatch,
        bool alsoOnReversal,
        string? failureMessage,
        LocalParallelizer parallelizer)
    {
        Regex parsed;
        try
        {
            parsed = new Regex(regEx, RegexOptions.Multiline);
        }
        catch (ArgumentException ex)
        {
            throw StarlarkRt.Errorf("Regex '{0}' is invalid: {1}", regEx, ex.Message);
        }
        return new VerifyMatch(
            parsed, verifyNoMatch, alsoOnReversal, paths, failureMessage, parallelizer, location);
    }
}
